### PR TITLE
Read X11 keyboard from /etc/vconsole.conf

### LINKED ISSUE #24228: Request for localectl to write the X11 keymap to /etc/vconsole.conf
### Component

other

### Is your feature request related to a problem? Please describe

I hope I am describing this issue correctly, this came up on a draft merge request for Plymouth to make it so it doesn't rely on the kernel mode console for input, but evdev instead.

localectl writes the keymaps to the kernel console in /etc/vconsole.conf, but writes the value for the X11 keymaps in an X11 file, and it seems like it's smart enough to translate the keymap names, As I understand X11 (or user mode) keymap names are not one-to-one with the console keymap names

User mode processes that are not X11, and can't understand X11 config files easily might want a single place to refer to the default configured keymap that libxkbcommon understands.

Having the value saved in  /etc/vconsole.conf by localectl might be good because the file is already known, maybe as an additional key like `XKB_KEYMAP`

### Describe the solution you'd like

It might be hard since now the value will be in two places since one would have to win, (like if someone manually edits the value on vconsole.conf, and not the X11 file, and now the current default one has to be read) but writing the XKB keymap as a key such as `XKB_KEYMAP=foo` to vconsole.conf would allow things that are not Xorg, such as some Wayland servers, or Plymouth. (OK mostly Plymouth) 

Plymouth on distros that use dracut already have /etc/vconsole.conf in their initrd, so amending this file might be the best bet for it

### Describe alternatives you've considered

_No response_

### The systemd version you checked that didn't have the feature you are asking for

_No response_

### PR DESCRIPTION
**Pull request by Diegovsky**
Closes #24228

### CODE REVIEW

**Comment by bluca** [comment]
```
18:21:40 [   24.610007] testsuite-73.sh[828]: FAIL: 'X11 Layout: .unset.' not found in:
18:21:40 [   24.610709] testsuite-73.sh[828]: System Locale: LANG=C.UTF-8
18:21:40 [   24.613520] testsuite-73.sh[828]:     VC Keymap: us
18:21:40 [   24.614159] testsuite-73.sh[828]:    X11 Layout: us
18:21:40 [   24.614912] testsuite-73.sh[828]:     X11 Model: pc105+inet
18:21:40 [   24.615570] testsuite-73.sh[828]:   X11 Variant: intl
18:21:40 [   24.616213] testsuite-73.sh[828]:   X11 Options: terminate:ctrl_alt_bksp
```

**Comment by Diegovsky** [comment]
Sorry, I have no idea what is missing. GH CI says all tests passed so I'm confused

**Comment by halfline** [comment]
I think you click here:

Screenshot from 2022-11-18 13-24-35

which links to here:

https://autopkgtest.ubuntu.com/results/autopkgtest-jammy-upstream-systemd-ci-systemd-ci/jammy/amd64/s/systemd-upstream/20221116_214115_a208d@/log.gz

which shows
```
[ 3833.899612] testsuite-73.sh[33]: + rm -f /etc/X11/xorg.conf.d/00-keyboard.conf /etc/default/keyboard
[ 3833.908186] testsuite-73.sh[236]: ++ localectl
         Starting [0;1;39mLocale Service[0m...
[ 3834.010197] testsuite-73.sh[33]: + output='System Locale: LANG=de_DE.UTF-8
[ 3834.010492] testsuite-73.sh[33]:     VC Keymap: us
[ 3834.010757] testsuite-73.sh[33]:    X11 Layout: us
[ 3834.011057] testsuite-73.sh[33]:     X11 Model: pc105+inet
[ 3834.011321] testsuite-73.sh[33]:   X11 Variant: intl
[ 3834.011580] testsuite-73.sh[33]:   X11 Options: terminate:ctrl_alt_bksp'
[ 3834.011839] testsuite-73.sh[33]: + assert_in 'X11 Layout: .unset.' 'System Locale: LANG=de_DE.UTF-8
[ 3834.012099] testsuite-73.sh[33]:     VC Keymap: us
[ 3834.012357] testsuite-73.sh[33]:    X11 Layout: us
[ 3834.012614] testsuite-73.sh[33]:     X11 Model: pc105+inet
[ 3834.012883] testsuite-73.sh[33]:   X11 Variant: intl
[ 3834.013143] testsuite-73.sh[33]:   X11 Options: terminate:ctrl_alt_bksp'
[ 3834.014149] testsuite-73.sh[238]: + set +ex
[ 3834.014647] testsuite-73.sh[238]: FAIL: 'X11 Layout: .unset.' not found in:
[ 3834.014993] testsuite-73.sh[238]: System Locale: LANG=de_DE.UTF-8
[ 3834.015326] testsuite-73.sh[238]:     VC Keymap: us
[ 3834.015583] testsuite-73.sh[238]:    X11 Layout: us
[ 3834.015848] testsuite-73.sh[238]:     X11 Model: pc105+inet
[ 3834.016107] testsuite-73.sh[238]:   X11 Variant: intl
[ 3834.016400] testsuite-73.sh[238]:   X11 Options: terminate:ctrl_alt_bksp
```

Notice the `rm -f` at the top. It's trying to erase the X11 configuration to make sure console keymaps are converted into a passable X11 layouts (in case the user never had an X11 layout set).

After your change, the `rm -f /etc/X11/xorg.conf.d/00-keyboard.conf /etc/default/keyboard` is no longer sufficient to erase the X11 configuration.

**Comment by n3rdopolis** [comment]
Would he need to 

```
diff --git a/test/units/testsuite-73.sh b/test/units/testsuite-73.sh
index d96e9f4151..deda8b84d3 100755
--- a/test/units/testsuite-73.sh
+++ b/test/units/testsuite-73.sh
@@ -372,7 +372,7 @@ XKBMODEL=pc105+inet"
 
     # gets along without config file
     systemctl stop systemd-localed.service
-    rm -f /etc/X11/xorg.conf.d/00-keyboard.conf /etc/default/keyboard
+    rm -f /etc/X11/xorg.conf.d/00-keyboard.conf /etc/default/keyboard /etc/vconsole.conf
     output=$(localectl)
     assert_in "X11 Layout: n/a" "$output"
     assert_not_in "X11 Model:" "$output"
```
?

**Comment by halfline** [comment]
no that would get rid of the console config getting tested too. I think maybe

```diff 
 systemctl stop systemd-localed.service
 rm -f /etc/X11/xorg.conf.d/00-keyboard.conf /etc/default/keyboard
+ sed -ie '/^XKB_/d' /etc/vconsole.conf
 output=$(localectl)
```

or maybe just a canned one created from scratch? Not sure what the systemd maintainers think...

File: src/locale/localed-util.c
Comment (yuwata) [inline src/locale/localed-util.c]: Coding style:
- please add a space between `if` and `(`.
- please use implicit cast from pointer to boolean: `c->x11_layout != NULL` should be `c->x11_layout`. Though, in this case, `!isempty(c->x11_layout)` is better.
- our code does not set `errno`. Please check the return value `r`.
- use `if (r >= 0)` instead of `if (r == 0)` to check success, as the function may be changed to return positive value in the future.

So,
```c
if (r < 0)
        return r;
if (!isempty(c->x11_layout)) {
        log_debug("Found X11 keyboard data in /etc/vconsole.conf, skipping to read /etc/X11/xorg.conf.d/00-keyboard.conf");
        return 0;
}
```

**Review by yuwata — CHANGES_REQUESTED** [review]
`vconsole_write_data()` needs to be called in `method_set_x11_keyboard()`.

**Comment by Diegovsky** [comment]
Thank you everyone for the input, I'll implement the changes ASAP

File: src/locale/localed-util.c
Comment (Diegovsky) [reply src/locale/localed-util.c]: Noted, thanks for the pointers :+1: 

I was checking for `errno` to prevent the X11 layout from not being set if `vconsole.conf` does not exist and instead read the xorg config file. I should check `r` instead

Comment (Diegovsky) [reply src/locale/localed-util.c]: Also, since I call `vconsole_read_data` inside `x11_read_data`, should I call `vconsole_write_data` inside `x11_write_data`?

File: src/locale/localed-util.c
Comment (yuwata) [inline src/locale/localed-util.c:151]: The function `vconsole_read_data()` should not return `-ENOENT`.

File: src/locale/localed-util.c
Comment (yuwata) [inline src/locale/localed-util.c:155]: Please use 8ch indentation.

File: src/locale/localed.c
Comment (yuwata) [inline src/locale/localed.c:664]: This should be earlier than `x11_write_data()`, as now vconsole.conf has precedence over X configuration.

Also, `x11_convert_to_vconsole_and_emit()` below also writes vconsole.conf. Please try to convert earlier, and write vconsole.conf only once.

**Review by yuwata — CHANGES_REQUESTED** [review]

File: src/locale/localed-util.c
Comment (yuwata) [reply src/locale/localed-util.c]: > Also, since I call `vconsole_read_data` inside `x11_read_data`, should I call `vconsole_write_data` inside `x11_write_data`?

I do not think so, as the error should be propagated to the DBus client.

Comment (Diegovsky) [reply src/locale/localed-util.c]: ok

File: src/locale/localed-util.c
Comment (Diegovsky) [reply src/locale/localed-util.c:151]: I see. In that case, how can we check if the `/etc/vconsole.conf` file did not exist? It seems to me this code will skip reading the Xorg keyboard config file if vconsole does not exist

File: src/locale/localed-util.c
Comment (Diegovsky) [reply src/locale/localed-util.c:155]: sorry, my editor got messed up

File: src/locale/localed-util.c
Comment (halfline) [reply src/locale/localed-util.c:151]: @yuwata  I think you're saying `vconsole_read_data` won't return `-ENOENT` because it stats the file before trying to parse it and if the file isn't there, it returns success.

But that stat call is racy of course, so @Diegovsky is right, there is an edge case where `vconsole_read_data` can return `-ENOENT`. It's clear the intention of the code is for `vconsole_read_data` never to return `-ENOENT`, though, so maybe this pull request should ignore that problem ...

I guess we could also take the opportunity to just fix the race (e.g. call `xfopenat(AT_FDCWD, "/etc/vconsole.conf", ..., &f)` instead of `stat`, and pass `f` to `parse_env_file` instead of the filename, or i guess handle `-ENOENT` after `parse_env_file`)

Comment (Diegovsky) [reply src/locale/localed-util.c:151]: I like your solution, @halfline but it seems to require some substantial changes to systemd's codebase. Though, I can make a  function that uses a `FILE*` instead of a path and contains all the logic from `parse_env_file` and call it from `parse_env_file`. What do you think @yuwata ?

Comment (halfline) [reply src/locale/localed-util.c:151]: I think it's just swapping the `NULL` from the first argument to the second...The prototype is:

`int parse_env_file(FILE *f, const char *fname, ...)`

(or again, you could just drop the stat call, leave `parse_env_file` args the same, I suppose, and handle `-ENOENT` after `parse_env_file` returns instead)

Comment (halfline) [reply src/locale/localed-util.c:151]: @Diegovsky is this still on your radar ? I don't want to do a plymouth release until this pull request has landed, since I don't want to rely on undocumented features of the config file. (If you're too busy with school etc, no biggie, me or @n3rdopolis can probably pick it up and do the last couple bits, though it would probably have to be in a different pull request)

Comment (Diegovsky) [reply src/locale/localed-util.c:151]: Yeah, I'm very busy with college finals, sorry for taking so long to update :(

Edit: added both of you to the repo but feel free to open a new pull request if you want

File: src/locale/localed.c
Comment (n3rdopolis) [reply src/locale/localed.c:664]: so 
```
diff --git a/src/locale/localed.c b/src/locale/localed.c
index 209d36a0ed..132a61ddae 100644
--- a/src/locale/localed.c
+++ b/src/locale/localed.c
@@ -656,16 +656,16 @@ static int method_set_x11_keyboard(sd_bus_message *m, void *userdata, sd_bus_err
             free_and_strdup(&c->x11_options, options) < 0)
                 return -ENOMEM;
 
-        r = x11_write_data(c);
-        if (r < 0) {
-                log_error_errno(r, "Failed to set X11 keyboard layout: %m");
-                return sd_bus_error_set_errnof(error, r, "Failed to set X11 keyboard layout: %m");
-        }
         r = vconsole_write_data(c);
         if (r < 0) {
                 log_error_errno(r, "Failed to save X11 keyboard layout to /etc/vconsole.conf: %m");
                 return sd_bus_error_set_errnof(error, r, "Failed to save X11 keyboard layout to /etc/vconsole.conf: %m");
         }
+        r = x11_write_data(c);
+        if (r < 0) {
+                log_error_errno(r, "Failed to set X11 keyboard layout: %m");
+                return sd_bus_error_set_errnof(error, r, "Failed to set X11 keyboard layout: %m");
+        }
 
         log_info("Changed X11 keyboard layout to '%s' model '%s' variant '%s' options '%s'",
                  strempty(c->x11_layout),

```
?

**Comment by n3rdopolis** [comment]
> I am working on this based on @Diegovsky's work.

Thanks, let me know if you need to test, or need me to try to grant you permissions to this pull request

**Comment by yuwata** [comment]
Replaced by #25805.

**Comment by n3rdopolis** [comment]
This can be closed I guess since https://github.com/systemd/systemd/pull/25805 replaces this, and is merged

**Comment by poettering** [comment]
See → https://github.com/systemd/systemd/pull/26182, which renames the variables.

**Pull request closed without merging**
